### ISSUE TITLE
Any updates on docking?

### ISSUE THREAD

**Issue by Qlii256** [swistakm/pyimgui#227]
This project seems to be "dead", but I'm really loving it! I'm still waiting for docking and even dragging windows out of the actual window. That would be so cool to have that and I was wondering if there's any updates on this.

**Comment by KinoxKlark** [comment]
Unfortunately, the development is a little bit slowed down due to lake of time from its main author who has the right access to the repository. Currently, we are working on an upgrade to the latest Dear ImGui in #192, more than a simple upgrade, this branch is also a large extension of the number of mapped functions and functionalities, with various bug fixes.

The plan, as I understand it, is to use this upgrade as the basis of the next major release. But it is unclear when this will be possible since it depends on the owner @swistakm and its availability.

Concerning docking, I am willing to work on it in the future but the priority is given to the upgrade and the future release. The question of matching the docking branch of dear imgui has been raised and no clear plan for now. It may be implemented as a separate branch as in the cpp library. But it raises some questions about PIP domains or the possibility of making it as an extension.

I know that this is a bit of an unclear answer, I hope that we'll be able to move forward with the development in a near future but there is a lot of uncertainties. Hope this is still a little bit helpful :)

**Comment by Qlii256** [comment]
> I've just tried out the docking branch of imgui-cpp within pyimgui version-2.0 branch. Not too many changes required to make it build. The docking demo works fine, but I haven't tried anything else.

Is there a way I can try this out? I'd really love to have docking available with my GLFW application.

**Comment by KinoxKlark** [comment]
Hello, you can install from #242, but I don't think there is any mapping yet. I plan to merge this PR in a `docking` branch on which we will be able to provide mapping for those functionalities. Not sure if this will be merged as part of the official pyimgui, if it will have its own pypi package, or if it will stay for manual installation. We'll see later.

**Comment by Qlii256** [comment]
By mapping, you mean methods/functions I can use to enable the docking space? I'm very eager to finally have docking with my Python projects. I did try DearPyGui, and it does work in a more pythonic way, but it does not allow me to have my own viewport to which it renders so. And there's no integrations with GLFW or anything else.

**Comment by KinoxKlark** [comment]
> By mapping, you mean methods/functions I can use to enable the docking space? 

Yes exactly, this PR does only the bare minimum to compile `pyimgui` using the docking branch of DearImGui. But there are no functions on the python side of `pyimgui` that can be called. 

>I'm very eager to finally have docking with my Python projects. I did try DearPyGui, and it does work in a more pythonic way, but it does not allow me to have my own viewport to which it renders so. And there's no integrations with GLFW or anything else.

I am aware that these functionalities are wanted by many. Unfortunately, I, myself, do not have a lot of free time to invest in that project recently and there are still more urgent matters with the coming 2.0 release which is a prerequisite to the docking branch. I hope to be able to work on it soon after but it may have to wait a little more.

Of course, anyone is welcome to start to tackle the task. I'll try to create a development branch for this functionality this week so that motivated contributors could propose some PR to it.

**Comment by rempelj** [comment]
Hey @KinoxKlark, first of all thanks for taking on this package. I have been following it for a while and I'm really glad it's able to stay alive!

You mentioned above that you are not sure if docking will be merged as part of the official pyimgui, if it will have its own pypi package, or if it will stay for manual installation. 

It would be extremely useful to have the docking branch available via pypi so that docking can be a dependency in `setup.py` or installed with `pip install`. 

I think we should have an issue to track "docking available via pypi" as the ultimate goal for docking. What do you think?

**Comment by KinoxKlark** [comment]
Yes, I agree with you, having a package to refer to in dependencies seems to be a must-have. I guess the question is, should we use another pypi package like `pyimgui-docking`, with its own `[glfw]`, `[...]` backend dependencies? Or should we propose the docking functionalities directly in the `pyimgui` package, etc? Anyway, this is a question for the future since the functionality is not at all ready to be used. I do not remember if I announced it somewhere but I created a `docking` branch for the ones that are motivated to map functionalities. (See https://github.com/pyimgui/pyimgui/tree/docking). I am not sure to understand what you mean by "I think we should have an open issue to track "docking available via pypi" as the ultimate goal for docking."?

For now, I need to focus on releasing 2.0, which is a prerequisite to the docking functionality anyway. If anyone wants to work on that branch I will obviously consider any PR. I'll see what I can do myself when 2.0 will be out and when I'll have more time to spend on this project.

**Comment by pvallet** [comment]
@KinoxKlark why do you want to have docking as a separate package? From a user’s perspective it does seem more natural to have everything installed as part of the same package (pyimgui provides bindings for everything in dear imgui), but maybe I missed something? If it’s because the feature is too experimental, there might be another way to signify it to the user from inside the package.

**Comment by rempelj** [comment]
Thanks for the reply! I agree with you that some of these questions don't need to be answered right now. What I mean by "an open issue to track docking available via pypi as the ultimate goal for docking." is that we should have an issue here on github to track the progress of the docking release and discuss these details. Maybe we could just repurpose this ticket and rename it to "Docking available via pypi" or "Release docking on pypi" for example.

**Comment by learn-more** [comment]
> Thanks for the reply! I agree with you that some of these questions don't need to be answered right now, but it would be nice to have a clear place to discuss our options and track the progress towards docking.
> 
> What I mean by "an open issue to track docking available via pypi as the ultimate goal for docking." is that we should have a [#todo](https://github.com/pyimgui/pyimgui/labels/todo) issue here on github to track the progress of the docking release and discuss these details.
> 
> Maybe we could just repurpose this issue and rename it to "Docking available via pypi" or "Release docking on pypi" for example and add the "todo" label.

I went ahead and created #259 to track this progress.
@KinoxKlark I assume you are able to edit this issue?
